File: app/src/main/java/nl/rsdt/japp/application/navigation/FragmentNavigationManager.java
package nl.rsdt.japp.application.navigation;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v7.app.ActionBar;


import java.util.ArrayList;

import nl.rsdt.japp.R;
import nl.rsdt.japp.application.activities.MainActivity;
import nl.rsdt.japp.application.fragments.AboutFragment;
import nl.rsdt.japp.application.fragments.HomeFragment;
import nl.rsdt.japp.application.fragments.JappMapFragment;
import nl.rsdt.japp.application.fragments.JappPreferenceFragment;
import nl.rsdt.japp.application.fragments.CarFragment;
import nl.rsdt.japp.jotial.maps.wrapper.IJotiMap;

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 8-7-2016
 * Description...
 */
public class FragmentNavigationManager {

    public static final String FRAGMENT_HOME = "FRAGMENT_HOME";

    public static final String FRAGMENT_MAP = "FRAGMENT_MAP";

    public static final String FRAGMENT_SETTINGS = "FRAGMENT_SETTINGS";

    public static final String FRAGMENT_ABOUT = "FRAGMENT_ABOUT";

    public static final String FRAGMENT_CAR = "FRAGMENT_CAR";

    private static final String BUNDLE_KEY_FRAGMENT = "BUNDLE_KEY_FRAGMENT";

    private ArrayList<String> backstack = new ArrayList<>();

    private NavigationView nView;

    private ActionBar actionbar;

    private FragmentManager manager;

    private String currentFragmentTag = "";

    private HomeFragment homeFragment;

    private JappMapFragment mapFragment;

    private CarFragment carFragment;

    private JappPreferenceFragment preferenceFragment;

    private AboutFragment aboutFragment;

    public Fragment getFragment(String tag) {
        switch(tag) {
            case FRAGMENT_HOME:
                return homeFragment;
            case FRAGMENT_CAR:
                return carFragment;
            case FRAGMENT_MAP:
                return mapFragment;
            case FRAGMENT_SETTINGS:
                return preferenceFragment;
            case FRAGMENT_ABOUT:
                return aboutFragment;
        }
        return null;
    }

    public boolean hasBackStack() {
        return !backstack.isEmpty();
    }

    public void initialize(MainActivity mainActivity)
    {
        this.nView = (NavigationView) mainActivity.findViewById(R.id.nav_view);
        this.actionbar = mainActivity.getSupportActionBar();
        this.manager = mainActivity.getFragmentManager();
        setupFragments();
    }

    public void onSaveInstanceState(Bundle saveInstanceState)
    {
        if(!currentFragmentTag.isEmpty())
        {
            saveInstanceState.putString(BUNDLE_KEY_FRAGMENT, currentFragmentTag);
        }
    }

    public void onSavedInstance(Bundle savedInstanceState)
    {
        if(savedInstanceState != null)
        {
            String value = savedInstanceState.getString(BUNDLE_KEY_FRAGMENT, currentFragmentTag);
            switchTo(value);
        }
    }

    public void switchTo(String fragment) {
       switchTo(fragment, true);
    }

    private void switchTo(String fragment, boolean addToStack) {
        if(currentFragmentTag == null || currentFragmentTag.equals(fragment)) return;

        FragmentTransaction ft = manager.beginTransaction();

        switch (fragment)
        {
            case FRAGMENT_HOME:
                ft.show(homeFragment);
                break;
            case FRAGMENT_MAP:
                ft.show(mapFragment);
                break;
            case FRAGMENT_CAR:
                ft.show(carFragment);
                break;
            case FRAGMENT_SETTINGS:
                ft.show(preferenceFragment);
                break;
            case FRAGMENT_ABOUT:
                ft.show(aboutFragment);
                break;
        }

        switch (currentFragmentTag)
        {
            case FRAGMENT_HOME:
                ft.hide(homeFragment);
                break;
            case FRAGMENT_MAP:
                ft.hide(mapFragment);
                break;
            case FRAGMENT_CAR:
                ft.hide(carFragment);
                break;
            case FRAGMENT_SETTINGS:
                ft.hide(preferenceFragment);
                break;
            case FRAGMENT_ABOUT:
                ft.hide(aboutFragment);
                break;
        }

        ft.commit();
        if(addToStack) {
            backstack.add(currentFragmentTag);
        }
        currentFragmentTag = fragment;
        updateToolbarTitle();
        updateCheckedState();
    }

    public void setupMap(IJotiMap.OnMapReadyCallback callback)
    {
        if (mapFragment == null) {//// TODO: 09/08/17 dit is nooit anders krijg je een runtime exception bij de vorige if statement
            mapFragment = (JappMapFragment) manager.findFragmentByTag(FRAGMENT_MAP);
            mapFragment.getMapAsync(callback);
        } else {
            mapFragment.getMapAsync(callback);
        }
    }

    public void onBackPressed() {
        if(!backstack.isEmpty()) {
            int last = backstack.size() - 1;
            switchTo(backstack.get(last), false);
            backstack.remove(last);
        }
    }

    private void updateToolbarTitle() {
        if(currentFragmentTag != null && actionbar != null) {
            switch (currentFragmentTag)
            {
                case FRAGMENT_HOME:
                    actionbar.setTitle(R.string.home);
                    break;
                case FRAGMENT_MAP:
                    actionbar.setTitle(R.string.map);
                    break;
                case FRAGMENT_CAR:
                    actionbar.setTitle(R.string.car);
                    break;
                case FRAGMENT_SETTINGS:
                    actionbar.setTitle(R.string.settings);
                    break;
                case FRAGMENT_ABOUT:
                    actionbar.setTitle(R.string.about);
                    break;
            }
        }
    }

    private void updateCheckedState() {

        switch(currentFragmentTag)
        {
            case FRAGMENT_HOME:
                nView.getMenu().findItem(R.id.nav_home).setChecked(true);
                break;
            case FRAGMENT_MAP:
                nView.getMenu().findItem(R.id.nav_map).setChecked(true);
                break;
            case FRAGMENT_CAR:
                nView.getMenu().findItem(R.id.nav_car).setChecked(true);
                break;
            case FRAGMENT_SETTINGS:
                nView.getMenu().findItem(R.id.nav_settings).setChecked(true);
                break;
            case FRAGMENT_ABOUT:
                nView.getMenu().findItem(R.id.nav_about).setChecked(true);
                break;
        }
    }


    private void setupFragments()
    {
        FragmentTransaction ft = manager.beginTransaction();

        homeFragment = (HomeFragment) manager.findFragmentByTag(HomeFragment.TAG);
        if(homeFragment == null)
        {
            homeFragment = new HomeFragment();
            ft.add(R.id.container, homeFragment, HomeFragment.TAG);
        }
        ft.hide(homeFragment);

        carFragment = (CarFragment) manager.findFragmentByTag(CarFragment.TAG);
        if(carFragment == null)
        {
            carFragment = new CarFragment();
            ft.add(R.id.container, carFragment, CarFragment.TAG);
        }
        ft.hide(carFragment);

        mapFragment = (JappMapFragment) manager.findFragmentByTag(JappMapFragment.TAG);
        if(mapFragment == null)
        {
            mapFragment = new JappMapFragment();
            ft.add(R.id.container, mapFragment, JappMapFragment.TAG);
        }
        ft.show(mapFragment);
        currentFragmentTag = FRAGMENT_MAP;
        updateToolbarTitle();
        updateCheckedState();

        preferenceFragment  = (JappPreferenceFragment)manager.findFragmentByTag(JappPreferenceFragment.TAG);
        if(preferenceFragment == null)
        {
            preferenceFragment = new JappPreferenceFragment();
            ft.add(R.id.container, preferenceFragment, JappPreferenceFragment.TAG);
        }
        ft.hide(preferenceFragment);

        aboutFragment = (AboutFragment)manager.findFragmentByTag(AboutFragment.TAG);
        if(aboutFragment == null)
        {
            aboutFragment = new AboutFragment();
            ft.add(R.id.container, aboutFragment, AboutFragment.TAG);
        }
        ft.hide(aboutFragment);

        ft.commit();
    }

    public void onDestroy()
    {
        if(manager != null)
        {
            manager = null;
        }

        if(homeFragment != null)
        {
            homeFragment = null;
        }

        if(mapFragment != null)
        {
            mapFragment = null;
        }
        if(carFragment != null)
        {
            carFragment = null;
        }
        if(preferenceFragment != null)
        {
            preferenceFragment = null;
        }

        if(aboutFragment != null)
        {
            aboutFragment = null;
        }


    }


}

File: app/src/main/java/nl/rsdt/japp/application/navigation/FragmentNavigationManager.kt
package nl.rsdt.japp.application.navigation

import android.app.Fragment
import android.app.FragmentManager
import android.os.Bundle
import androidx.appcompat.app.ActionBar
import com.google.android.material.navigation.NavigationView
import nl.rsdt.japp.R
import nl.rsdt.japp.application.activities.MainActivity
import nl.rsdt.japp.application.fragments.*
import nl.rsdt.japp.jotial.maps.wrapper.IJotiMap
import java.util.*

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 8-7-2016
 * Description...
 */
open class FragmentNavigationManager {

    private val backstack = ArrayList<String>()

    private var nView: NavigationView? = null

    private var actionbar: ActionBar? = null

    private var manager: FragmentManager? = null

    private var currentFragmentTag: String = ""

    private var homeFragment: HomeFragment? = null

    private var mapFragment: JappMapFragment? = null

    private var carFragment: CarFragment? = null

    private var preferenceFragment: JappPreferenceFragment? = null

    private var aboutFragment: AboutFragment? = null

    private var helpFragment: HelpFragment? = null

    fun getFragment(tag: String): Fragment? {
        when (tag) {
            FRAGMENT_HOME -> return homeFragment
            FRAGMENT_CAR -> return carFragment
            FRAGMENT_MAP -> return mapFragment
            FRAGMENT_SETTINGS -> return preferenceFragment
            FRAGMENT_ABOUT -> return aboutFragment
            FRAGMENT_HELP -> return helpFragment
        }
        return null
    }

    fun hasBackStack(): Boolean {
        return backstack.isNotEmpty()
    }

    open fun initialize(mainActivity: MainActivity) {
        this.nView = mainActivity.findViewById(R.id.nav_view)
        this.actionbar = mainActivity.supportActionBar
        this.manager = mainActivity.fragmentManager
        setupFragments()
    }

    fun onSaveInstanceState(saveInstanceState: Bundle?) {
        if (!currentFragmentTag.isEmpty()) {
            saveInstanceState?.putString(BUNDLE_KEY_FRAGMENT, currentFragmentTag)
        }
    }

    fun onSavedInstance(savedInstanceState: Bundle?) {
        if (savedInstanceState != null) {
            val value = savedInstanceState.getString(BUNDLE_KEY_FRAGMENT, currentFragmentTag)
            switchTo(value)
        }
    }

    fun switchTo(fragment: String) {
        switchTo(fragment, true)
    }

    private fun switchTo(fragment: String, addToStack: Boolean) {
        if (currentFragmentTag == fragment) return

        val ft = manager?.beginTransaction()

        when (fragment) {
            FRAGMENT_HOME -> ft?.show(homeFragment)
            FRAGMENT_MAP -> ft?.show(mapFragment)
            FRAGMENT_CAR -> ft?.show(carFragment)
            FRAGMENT_SETTINGS -> ft?.show(preferenceFragment)
            FRAGMENT_ABOUT -> ft?.show(aboutFragment)
            FRAGMENT_HELP -> ft?.show(helpFragment)
        }

        when (currentFragmentTag) {
            FRAGMENT_HOME -> ft?.hide(homeFragment)
            FRAGMENT_MAP -> ft?.hide(mapFragment)
            FRAGMENT_CAR -> ft?.hide(carFragment)
            FRAGMENT_SETTINGS -> ft?.hide(preferenceFragment)
            FRAGMENT_ABOUT -> ft?.hide(aboutFragment)
            FRAGMENT_HELP -> ft?.hide(helpFragment)
        }

        ft?.commit()
        if (addToStack) {
            backstack.add(currentFragmentTag)
        }
        currentFragmentTag = fragment
        updateToolbarTitle()
        updateCheckedState()
    }

    fun setupMap(callback: IJotiMap.OnMapReadyCallback) {
        if (mapFragment == null) {//// TODO: 09/08/17 dit is nooit anders krijg je een runtime exception bij de vorige if statement
            mapFragment = manager?.findFragmentByTag(FRAGMENT_MAP) as JappMapFragment
        }

        mapFragment?.getMapAsync(callback)

    }

    fun onBackPressed() {
        if (backstack.isNotEmpty()) {
            val last = backstack.size - 1
            switchTo(backstack[last], false)
            backstack.removeAt(last)
        }
    }

    private fun updateToolbarTitle() {
        if (actionbar != null) {
            when (currentFragmentTag) {
                FRAGMENT_HOME -> actionbar?.setTitle(R.string.home)
                FRAGMENT_MAP -> actionbar?.setTitle(R.string.map)
                FRAGMENT_CAR -> actionbar?.setTitle(R.string.car)
                FRAGMENT_SETTINGS -> actionbar?.setTitle(R.string.settings)
                FRAGMENT_ABOUT -> actionbar?.setTitle(R.string.about)
                FRAGMENT_HELP -> actionbar?.setTitle(R.string.help)
            }
        }
    }

    private fun updateCheckedState() {

        when (currentFragmentTag) {
            FRAGMENT_HOME -> nView?.menu?.findItem(R.id.nav_home)?.isChecked = true
            FRAGMENT_MAP -> nView?.menu?.findItem(R.id.nav_map)?.isChecked = true
            FRAGMENT_CAR -> nView?.menu?.findItem(R.id.nav_car)?.isChecked = true
            FRAGMENT_SETTINGS -> nView?.menu?.findItem(R.id.nav_settings)?.isChecked = true
            FRAGMENT_ABOUT -> nView?.menu?.findItem(R.id.nav_about)?.isChecked = true
            FRAGMENT_HELP -> nView?.menu?.findItem(R.id.nav_help)?.isChecked = true
        }
    }


    private fun setupFragments() {
        val ft = manager?.beginTransaction()

        homeFragment = manager?.findFragmentByTag(HomeFragment.TAG) as HomeFragment?
        if (homeFragment == null) {
            homeFragment = HomeFragment()
            ft?.add(R.id.container, homeFragment, HomeFragment.TAG)
        }
        ft?.hide(homeFragment)

        carFragment = manager?.findFragmentByTag(CarFragment.TAG) as CarFragment?
        if (carFragment == null) {
            carFragment = CarFragment()
            ft?.add(R.id.container, carFragment, CarFragment.TAG)
        }
        ft?.hide(carFragment)

        mapFragment = manager?.findFragmentByTag(JappMapFragment.TAG) as JappMapFragment?
        if (mapFragment == null) {
            mapFragment = JappMapFragment()
            ft?.add(R.id.container, mapFragment, JappMapFragment.TAG)
        }
        ft?.show(mapFragment)
        currentFragmentTag = FRAGMENT_MAP
        updateToolbarTitle()
        updateCheckedState()

        preferenceFragment = manager?.findFragmentByTag(JappPreferenceFragment.TAG) as JappPreferenceFragment?
        if (preferenceFragment == null) {
            preferenceFragment = JappPreferenceFragment()
            ft?.add(R.id.container, preferenceFragment, JappPreferenceFragment.TAG)
        }
        ft?.hide(preferenceFragment)

        aboutFragment = manager?.findFragmentByTag(AboutFragment.TAG) as AboutFragment?
        if (aboutFragment == null) {
            aboutFragment = AboutFragment()
            ft?.add(R.id.container, aboutFragment, AboutFragment.TAG)
        }
        ft?.hide(aboutFragment)

        helpFragment = manager?.findFragmentByTag(HelpFragment.TAG) as HelpFragment?
        if (helpFragment == null) {
            helpFragment = HelpFragment()
            ft?.add(R.id.container, helpFragment, HelpFragment.TAG)
        }
        ft?.hide(helpFragment)

        ft?.commit()
    }

    open fun onDestroy() {
        if (manager != null) {
            manager = null
        }

        if (homeFragment != null) {
            homeFragment = null
        }

        if (mapFragment != null) {
            mapFragment = null
        }
        if (carFragment != null) {
            carFragment = null
        }
        if (preferenceFragment != null) {
            preferenceFragment = null
        }

        if (aboutFragment != null) {
            aboutFragment = null
        }
        if (helpFragment != null) {
            helpFragment = null
        }


    }

    companion object {

        val FRAGMENT_HOME = "FRAGMENT_HOME"

        val FRAGMENT_MAP = "FRAGMENT_MAP"

        val FRAGMENT_SETTINGS = "FRAGMENT_SETTINGS"

        val FRAGMENT_ABOUT = "FRAGMENT_ABOUT"

        val FRAGMENT_CAR = "FRAGMENT_CAR"

        val FRAGMENT_HELP = "FRAGMENT_HELP"

        private val BUNDLE_KEY_FRAGMENT = "BUNDLE_KEY_FRAGMENT"
    }


}
